pkg/policy/verifier.go

```diff
@@ -0,0 +1,210 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"github.com/sigstore/policy-controller/pkg/apis/config"
+	"github.com/sigstore/policy-controller/pkg/webhook"
+	webhookcip "github.com/sigstore/policy-controller/pkg/webhook/clusterimagepolicy"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+)
+
+// Verifier is the interface for checking that a given image digest satisfies
+// the policies backing this interface.
+type Verifier interface {
+	// Verify checks that the provided reference satisfies the backing policies.
+	Verify(context.Context, name.Reference, authn.Keychain) error
+}
+
+// WarningWriter is used to surface warning messages in a manner that
+// is customizable by callers that's suitable for their execution
+// environment.
+type WarningWriter func(string, ...interface{})
+
+// Compile turns a Verification into an executable Verifier.
+// Any compilation errors are returned here.
+func Compile(ctx context.Context, v Verification, ww WarningWriter) (Verifier, error) {
+	if err := v.Validate(ctx); err != nil {
+		return nil, err
+	}
+
+	ipc, err := gather(ctx, v, ww)
+	if err != nil {
+		// This should never hit for validated policies.
+		return nil, err
+	}
+
+	return &impl{
+		verification: v,
+		ipc:          ipc,
+		ww:           ww,
+	}, nil
+}
+
+func gather(ctx context.Context, v Verification, ww WarningWriter) (*config.ImagePolicyConfig, error) {
+	pol := *v.Policies
+	ipc := &config.ImagePolicyConfig{
+		Policies: make(map[string]webhookcip.ClusterImagePolicy, len(pol)),
+	}
+
+	for i, p := range pol {
+		var content string
+		switch {
+		case p.Data != "":
+			content = p.Data
+
+		case p.Path != "":
+			raw, err := os.ReadFile(p.Path)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		case p.URL != "":
+			resp, err := http.Get(p.URL)
+			if err != nil {
+				return nil, err
+			}
+			defer resp.Body.Close()
+			raw, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, err
+			}
+			content = string(raw)
+
+		default:
+			// This should never happen for a validated policy.
+			return nil, fmt.Errorf("unsupported policy shape: %v", p)
+		}
+
+		l, warns, err := ParseClusterImagePolicies(ctx, content)
```

Comment:
Yeah, this is kind of what I was talking about in the 'Validate' function :) But I'm wondering why we parse them above and throw away the results (in the Validate function). Given the comment below on L107 about compilation, wondering if it would make sense to to rename Validate to Compile and have it return the parsed/fetched policies instead of redoing that work here again, and any changes have to be put in both places.

Reply:
Or put another way, wonder if there's any reason to separate Validate / Compile into separate exposed functions, or just drop Validate and have Compile alone. Just wondering since Validate is public method, it's a bit confusing since Compile already calls it and does double the work right after calling Validate. The only thing it seems to do in addition is to convert to evaluatable policies. Also since Validate does not check for duplicate policies, seems better validation to just call Compile.

Reply:
There wasn't as much duplication when I originally went with this shape because I only had `Data`! :)

Check out the way I refactored this in the next push (tl;dr private `fetch` method to share the logic to produce `(string, error)`).

I'd sort of like to keep the `Validate` function to drop our original version of the same method downstream (as I mentioned in the `Secret` comment, but as I said there, I don't care that much, just a nice to have).